File: Go/day2/day2.go
package main

import (
	"os"
	"strconv"
	"strings"
)

const (
	MAX_RED   = 12
	MAX_GREEN = 13
	MAX_BLUE  = 14
)

type Color string

const (
	Red   Color = "red"
	Green Color = "green"
	Blue  Color = "blue"
)

type Game struct {
	id   string
	sets []Set
}

func (g Game) is_possible() bool {
	for _, set := range g.sets {
		if !set.is_possible() {
			return false
		}
	}

	return true
}

type Set struct {
	red   int
	green int
	blue  int
}

func (s Set) is_possible() bool {
	return s.red <= MAX_RED && s.green <= MAX_GREEN && s.blue <= MAX_BLUE
}

func parse_line(line string) Game {
	sets := make([]Set, 0)

	game_and_sets := strings.SplitN(line, ":", 2)

	game_as_string := game_and_sets[0]
	sets_as_strings := strings.Split(game_and_sets[1], ";")

	id := strings.SplitN(game_as_string, " ", 2)[1]

	for _, set_string := range sets_as_strings {
		red := 0
		green := 0
		blue := 0

		split_set := strings.Split(set_string, ", ")

		for _, color := range split_set {
			color = strings.TrimSpace(color)
			color_split := strings.Split(color, " ")

			number, _ := strconv.Atoi(color_split[0])
			color_name := color_split[1]

			switch Color(color_name) {
			case Red:
				red = number
			case Green:
				green = number
			case Blue:
				blue = number
			}

		}

		set := Set{red, green, blue}

		sets = append(sets, set)
	}

	return Game{id, sets}
}

func solve_part_one(file_name string) int {
	contents_as_bytes, _ := os.ReadFile(file_name)

	contents := string(contents_as_bytes)

	total := 0
	for _, line := range strings.Split(contents, "\n") {
		if line == "" {
			continue
		}

		game := parse_line(line)

		if game.is_possible() {
			id_as_int, _ := strconv.Atoi(game.id)

			total += id_as_int
		}
	}

	return total
}

func main() {
	println(solve_part_one("input.txt"))
}

File: Haskell/day2/day2.hs
import Distribution.Utils.String (trim)

main = do
  contents <- readFile "input.txt"
  print $ solvePartOne contents
  print $ solvePartTwo contents

data Set = Set {red :: Int, green :: Int, blue :: Int} deriving (Show)

data Game = Game {idx :: Int, sets :: [Set]} deriving (Show)

split :: Char -> String -> [String]
split = split' [] []

split' :: [String] -> String -> Char -> String -> [String]
split' l current c [] = l ++ [current]
split' l current c (y : ys) = if c == y then split' (l ++ [current]) [] c ys else split' l (current ++ [y]) c ys

parseLine :: String -> Game
parseLine x = case separateId x of
  (i, s) -> Game {idx = i, sets = map getSet (split ';' s)}

separateId :: String -> (Int, String)
separateId x = case split ':' x of
  [a, b] -> case split ' ' a of
    [d, e] -> (read e :: Int, b)
    _ -> error "could not parse id"
  _ -> error "could not separate id"

getSet :: String -> Set
getSet s = getSet' (Set 0 0 0) (split ',' s)

getSet' :: Set -> [String] -> Set
getSet' set [] = set
getSet' set (x : xs) = case split ' ' (trim x) of
  [a, "red"] -> getSet' (Set {red = read (trim a) :: Int, green = green set, blue = blue set}) xs
  [a, "green"] -> getSet' (Set {red = red set, green = read (trim a) :: Int, blue = blue set}) xs
  [a, "blue"] -> getSet' (Set {red = red set, green = green set, blue = read (trim a)}) xs
  _ -> set

isPossible :: Set -> Bool
isPossible (Set {red, green, blue}) = red <= 12 && green <= 13 && blue <= 14

isPossibleGame :: Game -> Bool
isPossibleGame (Game {sets}) = all isPossible sets

getMaxSet :: Game -> Set
getMaxSet (Game {sets}) = Set {red = foldl max 0 (map red sets), green = foldl max 0 (map green sets), blue = foldl max 0 (map blue sets)}

getPower :: Set -> Int
getPower (Set x y z) = x * y * z

solvePartOne :: String -> Int
solvePartOne contents = sum $ map idx (filter isPossibleGame (map parseLine (lines contents)))

solvePartTwo :: String -> Int
solvePartTwo contents = sum $ map (getPower . getMaxSet . parseLine) (lines contents)